### PR TITLE
[Precision Depth Alignment] paddle.tanh reverse gradient calculation and Torch accuracy alignment.

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category

Operator Mechanism

### PR Types

Bug fixes

### Description

**主要修改点如下：**
1. 增加对 float16 类型的分支处理，其它类型走原逻辑不变。
2. 在 float16 分支中使用 CUDA half ：将 out 显式转换为半精度，用 __hmul 计算 out^2，显式“先乘”，避免被融合为 FMA。
3. 非 float16 类型保持原实现：dout * (one - out * out)。

**测试结果：**
1. paddle.nn.functional.tanh 全部与torch对齐（共120个case）；
2. paddle.tanh 全部与torch对齐（共142个case）；

pcard-67164

### CODE REVIEW

File: paddle/phi/kernels/funcs/activation_functor.h [hunk 1]
```diff
@@ -4483,7 +4483,13 @@ struct CudaTanhGradFunctor : public BaseActivationFunctor<T> {
 
   // dx = dout * (1 - out^2)
   __device__ __forceinline__ T operator()(const T dout, const T out) const {
-    return dout * (one - out * out);
+    if constexpr (std::is_same<T, phi::float16>::value) {
+      __half out_half = __float2half_rn(static_cast<float>(out));
+      __half tmp_half = __hmul(out_half, out_half);
+      return dout * (one - static_cast<T>(__half2float(tmp_half)));
```
Comment (A-nnonymous) [inline paddle/phi/kernels/funcs/activation_functor.h:4489]: 进了这个if， T应该就是float16了，为什么还要cast to float然后static cast回来，接着做减法？

Comment (zhengshengning) [reply paddle/phi/kernels/funcs/activation_functor.h:4489]: 因为没找到float16可以直接转_half的，所以先转成float

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by A-nnonymous — APPROVED** [review]
LGTM
